### PR TITLE
MONGOCRYPT-272 support createDataKey with GCP

### PR DESCRIPTION
**Pull request by kevinAlbs**

### CODE REVIEW

File: kms-message/src/kms_gcp_request.c [hunk 1]
```diff
@@ -84,11 +84,12 @@ kms_gcp_request_oauth_new (const char *host,
    req->crypto.sign_rsaes_pkcs1_v1_5 = kms_sign_rsaes_pkcs1_v1_5;
    if (opt->crypto.sign_rsaes_pkcs1_v1_5) {
       req->crypto.sign_rsaes_pkcs1_v1_5 = opt->crypto.sign_rsaes_pkcs1_v1_5;
+      req->crypto.sign_ctx = opt->crypto.sign_ctx;
```
Comment (kevinAlbs) [inline kms-message/src/kms_gcp_request.c:87]: This was a bug discovered when testing the callbacks. The signing context was not being copied from the `kms_request_opt` to the `kms_request_t`.

File: src/mongocrypt-kms-ctx.c [hunk 2]
```diff
@@ -438,10 +439,11 @@ _ctx_done_aws (mongocrypt_kms_ctx_t *kms, const char *json_field)
    return ret;
 }
 
-/* An Azure oauth KMS context has received full response. Parse out the bearer
+/* A Azure/GCP oauth KMS context has received full response. Parse out the
+ * bearer
```
Comment (kevinAlbs) [inline src/mongocrypt-kms-ctx.c:443]: I'll fix the weird newline break in the next commit.

File: src/mongocrypt-kms-ctx.c [hunk 3]
```diff
@@ -595,6 +599,114 @@ _ctx_done_azure_wrapkey_unwrapkey (mongocrypt_kms_ctx_t *kms)
    return ret;
 }
 
+/* An AWS KMS context has received full response. Parse out the result or error.
```
Comment (kevinAlbs) [inline src/mongocrypt-kms-ctx.c:602]: Oops, "AWS" should say "gcp" here. I'll fix in the next commit.

File: src/mongocrypt.c [hunk 4]
```diff
@@ -124,6 +124,17 @@ mongocrypt_new (void)
...
+
+   if (0 != _mongocrypt_once (_mongocrypt_do_init) ||
+       !(_native_crypto_initialized)) {
+      mongocrypt_status_t *status = crypt->status;
+
+      CLIENT_ERR ("failed to initialize");
+      /* Return crypt with failure status so caller can obtain error when
+       * calling mongocrypt_init */
+   }
+
```
Comment (kevinAlbs) [inline src/mongocrypt.c:137]: base64 decoding requires `_mongocrypt_do_init` to be called first. And now we base64 decode the GCP private key in `mongocrypt_setopt_kms_providers` before `mongocrypt_init`. So this now does the necessary global initialization when allocating a `mongocrypt_t`.

File: src/mongocrypt-ctx-private.h [hunk 5]
```diff
@@ -54,6 +63,7 @@ typedef struct __mongocrypt_ctx_opts_t {
    _mongocrypt_key_alt_name_t *key_alt_names;
    mongocrypt_encryption_algorithm_t algorithm;
    _mongocrypt_ctx_opts_azure_kek_t azure_kek;
+   _mongocrypt_ctx_opts_gcp_kek_t gcp_kek;
```
Comment (markbenvenuto) [inline src/mongocrypt-ctx-private.h:66]: We should use a union here for the azure and gcp options. You could a file a bug to do the same for AWS in the future.

File: src/mongocrypt-opts-private.h [hunk 6]
```diff
@@ -55,6 +62,9 @@ typedef struct {
    void *log_ctx;
    _mongocrypt_buffer_t schema_map;
    _mongocrypt_opts_kms_provider_azure_t kms_provider_azure;
```
Comment (markbenvenuto) [inline src/mongocrypt-opts-private.h:64]: Same note as with ctx, use a union for azure and gcp options

File: src/mongocrypt-private.h [hunk 7]
```diff
@@ -83,6 +83,7 @@ struct _mongocrypt_t {
    /* A counter, protected by mutex, for generating unique context ids */
    uint32_t ctx_counter;
    _mongocrypt_cache_oauth_t *cache_oauth_azure;
```
Comment (markbenvenuto) [inline src/mongocrypt-private.h:85]: use union

File: src/mongocrypt-kms-ctx.c [hunk 8]
```diff
@@ -595,6 +598,114 @@ _ctx_done_azure_wrapkey_unwrapkey (mongocrypt_kms_ctx_t *kms)
...
+         CLIENT_ERR ("Error in KMS response. HTTP status=%d", http_status);
+         goto fail;
+      }
+      /* GCP error responses include a JSON message, like { "message":
+       * "error", "code": <num> } */
+      bson_destroy (&body_bson);
+      if (!bson_init_from_json (&body_bson, body, body_len, &bson_error)) {
+         bson_init (&body_bson);
+      } else {
+         const char *msg = "";
```
Comment (markbenvenuto) [inline src/mongocrypt-kms-ctx.c:645]: Where do you use msg?

**Review by markbenvenuto — CHANGES_REQUESTED** [review]

File: src/mongocrypt-opts-private.h (hunk 6 above)
Comment (kevinAlbs) [reply src/mongocrypt-opts-private.h:64]: These store KMS provider options. A user can configure multiple concurrent KMS providers and, say, do auto encryption with keys encrypted by different providers. Though the KMS provider options for creating a datakey should be represented with a union.

File: src/mongocrypt-private.h (hunk 7 above)
Comment (kevinAlbs) [reply src/mongocrypt-private.h:85]: Similarly, we could be authenticating against Azure and GCP concurrently, so we need separate caches.

File: src/mongocrypt-ctx-private.h (hunk 5 above)
Comment (kevinAlbs) [reply src/mongocrypt-ctx-private.h]: Good point, I noted in https://jira.mongodb.org/browse/MONGOCRYPT-283 to do this for AWS.

**Review by markbenvenuto — APPROVED** [review]

**Pull request merged**